### PR TITLE
Gradle

### PR DESCRIPTION
**Pull request by AndreasMager**
TODO's:
- [x] gradle files
- [x] gradle setup
- [x] lint system
- [x] tar & zip generation in distribution
- [x] remove pom
- [x] remove assembly.xml
- [x] add bintray setup

### CODE REVIEW

**Comment by AndreasMager** [comment]
@andsel the osgi stuff don't have you an idea what is wrong?

File: build.gradle [hunk 1]
```diff
@@ -12,6 +12,8 @@ buildscript {
     }
   }
   dependencies {
+//        classpath group: 'com.harshil.gradle', name: 'felix-plugin', version: '1.3.1'
```
Comment (andsel) [inline build.gradle]: Why this commented and unused deps?

File: broker/build.gradle [hunk 2]
```diff
@@ -0,0 +1,62 @@
+
+description = 'Moquette - broker'
+
+//apply plugin: 'osgi'
+//apply plugin: 'com.harshild.felix-run'
+
+dependencies {
+    compile group: 'io.netty', name: 'netty-common', version: "${nettyVersion}"
```
Comment (andsel) [inline broker/build.gradle]: Instead of [version: "${nettyVersion}"] is it better [version: nettyVersion] ?

File: broker/build.gradle [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+apply plugin: 'maven'
+
+task createPom << {
+    pom {
+        project {
+        }
+    }.writeTo("build/pom.xml")
+}
+
+jar {
```
Comment (andsel) [inline broker/build.gradle]: Is it needed to create pom.xml files from Gradle?

File: broker/src/test/java/io/moquette/server/ServerIntegrationSSLClientAuthTest.java [hunk 4]
```diff
@@ -113,11 +114,18 @@
 
     IMqttClient m_client;
     MessageCollector m_callback;
+    static String backup;
 
     @BeforeClass
     public static void beforeTests() {
         String tmpDir = System.getProperty("java.io.tmpdir");
         s_dataStore = new MqttDefaultFilePersistence(tmpDir);
+        backup = System.getProperty("moquette.path");
```
Comment (andsel) [inline broker/src/test/java/io/moquette/server/ServerIntegrationSSLClientAuthTest.java]: This should be out of a patch that regards only the build script

File: broker/src/test/java/io/moquette/server/ServerLowlevelMessagesIntegrationTests.java [hunk 5]
```diff
@@ -81,10 +84,10 @@ public void elapseKeepAliveTime() throws InterruptedException {
          */
         m_client.sendMessage(connectMessage);
 
-        // wait 2 times the keepAlive
-        Thread.sleep(keepAlive * 2 * 1000);
-
-        assertTrue(m_client.isConnectionLost());
+        // wait 3 times the keepAlive
```
Comment (andsel) [inline broker/src/test/java/io/moquette/server/ServerLowlevelMessagesIntegrationTests.java]: Code modification should be out of scope of gradle build integration, it's better to separate in another PR

File: build.gradle [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    classpath 'net.ltgt.gradle:gradle-errorprone-plugin:0.0.8'
+    //classpath 'com.netflix.nebula:gradle-ospackage-plugin:2.2.6'
+    classpath 'com.github.jengelman.gradle.plugins:shadow:1.2.3'
+    //classpath 'net.ltgt.gradle:gradle-apt-plugin:0.10'
+  }
+}
+
+subprojects {
+
+  apply from: new File(rootDir, 'gradle/java.gradle')
```
Comment (andsel) [inline build.gradle]: Why include a separate gradle file?

File: distribution/build.gradle [hunk 7]
```diff
@@ -0,0 +1,59 @@
+
+description = 'Moquette - distribution'
+
+apply plugin: 'application'
+
+// adds task shadowJar - Create a combined JAR of project and runtime dependencies
+apply plugin: 'com.github.johnrengelman.shadow'
```
Comment (andsel) [inline distribution/build.gradle:7]: No shadow jar, we want a tag.gz with all the stuff packaged

**Comment by AndreasMager** [comment]
I have hacked the test. But there is an error:
```
ERROR: Bundle io.moquette.broker [16] Error starting file:/home/ama/dev/git/moquette/osgi_test/../broker/build/libs/moquette-broker-0.11-SNAPSHOT.jar (org.osgi.framework.BundleException: Unable to resolve io.moquette.broker [16](R 16.0): missing requirement [io.moquette.broker [16](R 16.0)] osgi.wiring.package; (osgi.wiring.package=com.bugsnag) Unresolved requirements: [[io.moquette.broker [16](R 16.0)] osgi.wiring.package; (osgi.wiring.package=com.bugsnag)])
org.osgi.framework.BundleException: Unable to resolve io.moquette.broker [16](R 16.0): missing requirement [io.moquette.broker [16](R 16.0)] osgi.wiring.package; (osgi.wiring.package=com.bugsnag) Unresolved requirements: [[io.moquette.broker [16](R 16.0)] osgi.wiring.package; (osgi.wiring.package=com.bugsnag)]
	at org.apache.felix.framework.Felix.resolveBundleRevision(Felix.java:4090)
	at org.apache.felix.framework.Felix.startBundle(Felix.java:2111)
	at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1365)
	at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:308)
	at java.lang.Thread.run(Thread.java:748)

```

**Comment by AndreasMager** [comment]
@andsel I am ready so far.
I don't know if the login bintray stuff works but it should. The Tar and ZIp upload is missing.
2 H2 test fails. They didn't run in maven due wrong folder layout.

**Comment by andsel** [comment]
@AndreasMager the maven build previously worked on maven (travis confirmed) why they now fails?

**Comment by AndreasMager** [comment]
Last master build 
```
[INFO] ------------------------------------------------------------------------
[INFO] Building Moquette - H2 storage 0.11-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-checkstyle-plugin:2.17:check (default) @ moquette-h2-storage ---
[INFO] Starting audit...
/home/travis/build/andsel/moquette/h2_storage/src/main/java/io/moquette/persistence/h2/H2MessagesStore.java:43:5: warning: Redundant 'public' modifier.
/home/travis/build/andsel/moquette/h2_storage/src/main/java/io/moquette/persistence/h2/H2MessagesStore.java:59:48: warning: '(' is followed by whitespace.
/home/travis/build/andsel/moquette/h2_storage/src/main/java/io/moquette/persistence/h2/H2PersistentQueue.java:0: warning: missing copyright header
/home/travis/build/andsel/moquette/h2_storage/src/main/java/io/moquette/persistence/h2/H2PersistentStore.java:44: warning: Line is longer than 120 characters (found 123).
/home/travis/build/andsel/moquette/h2_storage/src/main/java/io/moquette/persistence/h2/H2PersistentStore.java:53: warning: an empty line before }
/home/travis/build/andsel/moquette/h2_storage/src/main/java/io/moquette/persistence/h2/H2PersistentStore.java:80: warning: an empty line before }
/home/travis/build/andsel/moquette/h2_storage/src/main/java/io/moquette/persistence/h2/H2SessionsStore.java:143: warning: Line is longer than 120 characters (found 122).
Audit done.
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ moquette-h2-storage ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/travis/build/andsel/moquette/h2_storage/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ moquette-h2-storage ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-bundle-plugin:3.3.0:manifest (bundle-manifest) @ moquette-h2-storage ---
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ moquette-h2-storage ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/travis/build/andsel/moquette/h2_storage/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ moquette-h2-storage ---
[INFO] No sources to compile
[INFO] 
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ moquette-h2-storage ---
[INFO] No tests to run.
[INFO]      

 ```

The test folder was not in src. So maven didn't run it.

**Comment by andsel** [comment]
No test on moquette-h2-storage?! Uhhh not so good

**Comment by andsel** [comment]
I'm going to review in next days/weeks, moving to a new build system is important and I don't want t introduce things I can't manage. I've create some review request to open a dialog with on why certain choices, if I have any doubt can I post back here to you?

**Comment by AndreasMager** [comment]
Ok take your time. Feel free to ask. This gradle setup is the used in multiple big projects in our company.

File: mapdb_storage/build.gradle [hunk 8]
```diff
@@ -0,0 +1,7 @@
+
+description = 'Moquette - MapDB storage'
+dependencies {
+  compile project(':moquette-broker')
+  compile group: 'org.mapdb', name: 'mapdb', version:'1.0.8'
+  testCompile project(path:':moquette-broker', configuration: 'archives')
```
Comment (andsel) [inline mapdb_storage/build.gradle]: why is needed the configuration: 'archives'?

File: mapdb_storage/pom.xml [hunk 9]
```diff
@@ -1,129 +0,0 @@
...
-                <executions>
-                    <execution>
-                        <goals>
-                            <goal>test-jar</goal>
-                        </goals>
-                    </execution>
-                </executions>
-            </plugin>
-            <plugin>
-                <groupId>org.apache.felix</groupId>
```
Comment (andsel) [inline mapdb_storage/pom.xml:108]: I think that the corresponding stuff is not present in gradle file, how the manifest for OSGi is generated?

File: osgi_test/src/test/java/io/moquette/test/osgi/MoquetteOsgiTest.java [hunk 10]
```diff
@@ -35,15 +37,39 @@
...
+
+        assertTrue("broker buildfolder contains no jar", f.exists());
+
+        String url = "file:" + f.getAbsolutePath();
+        System.out.println(url);
         return options(
             junitBundles(),
-            mavenBundle("io.moquette", "moquette-broker")
+
+            //mavenBundle("com.bugsnag", "bugsnag")
```
Comment (andsel) [inline osgi_test/src/test/java/io/moquette/test/osgi/MoquetteOsgiTest.java]: Commented code has no meaning to stay in a PR

File: osgi_test/src/test/java/io/moquette/test/osgi/MoquetteOsgiTest.java [hunk 11]
```diff
@@ -35,15 +37,39 @@
...
 @RunWith(PaxExam.class)
 public class MoquetteOsgiTest {
+
     @Inject
     private BundleContext bundleContext;
 
+
     @Configuration
     public Option[] config() {
+        File libs = new File("../broker/build/libs/");
```
Comment (andsel) [inline osgi_test/src/test/java/io/moquette/test/osgi/MoquetteOsgiTest.java]: Is really needed these code changes, before it should have worked. I think is ugly on code going to a specific folder in build  path

**Comment by andsel** [comment]
Hi @AndreasMager 
I'm reviewing the PR and there some little changes or things to fix before squashing and include the PR:
- the PR is about switching from Maven to Gradle, so in my opinion no code changes or fixes external to the PR should be included.
- the maven build doesn't created uber Jar, so I don't understand the necessity to create shadow Jars
- I can't understand the meaning of gradle/test.gradle and what it does
- other things I described in the code reviews.

Thanks a lot for your effort and more importantly for your patient

 Andrea

File: broker/build.gradle (hunk 3 above)
Comment (AndreasMager) [reply broker/build.gradle]: The osgi stuff needs a pom in jar file. The maven jar has the pom file too. Without the pom the osgi test fails.

File: mapdb_storage/build.gradle (hunk 8 above)
Comment (AndreasMager) [reply mapdb_storage/build.gradle]: archives contains the output and sources from the test scope. mapdb test uses classes from broker/test

File: broker/src/test/java/io/moquette/server/ServerLowlevelMessagesIntegrationTests.java (hunk 5 above)
Comment (AndreasMager) [reply broker/src/test/java/io/moquette/server/ServerLowlevelMessagesIntegrationTests.java]: I will move the changes to a new PR. The juint testrunner setup differs in gradle. The position of the tests changed and the execution speed changed .

File: distribution/build.gradle (hunk 7 above)
Comment (AndreasMager) [reply distribution/build.gradle:7]: Ok will switch to multi jar build. I like uber jar ;)

File: osgi_test/src/test/java/io/moquette/test/osgi/MoquetteOsgiTest.java (hunk 10 above)
Comment (AndreasMager) [reply osgi_test/src/test/java/io/moquette/test/osgi/MoquetteOsgiTest.java]: I will remove the comments

File: broker/src/test/java/io/moquette/server/ServerLowlevelMessagesIntegrationTests.java (hunk 5 above)
Comment (andsel) [reply broker/src/test/java/io/moquette/server/ServerLowlevelMessagesIntegrationTests.java]: Oks, split the PR in two, one for the gradle and the other for the JUnit fix

File: osgi_test/src/test/java/io/moquette/test/osgi/MoquetteOsgiTest.java (hunk 11 above)
Comment (AndreasMager) [reply osgi_test/src/test/java/io/moquette/test/osgi/MoquetteOsgiTest.java]: I don't like this too, but i didn't find a better solution. With this changes the osgi test succeed the first time. The osgi test has failed on my machine in maven all the time.

Comment (andsel) [reply osgi_test/src/test/java/io/moquette/test/osgi/MoquetteOsgiTest.java]: But on Travis it was working, and also on my laptop. Does it depend by your setup?

File: mapdb_storage/pom.xml (hunk 9 above)
Comment (AndreasMager) [reply mapdb_storage/pom.xml:108]: I have added the osgi plugin and pom generation. 
See ad7db2e6152bb1825595a76fc3d61ae9fe906f04

File: osgi_test/src/test/java/io/moquette/test/osgi/MoquetteOsgiTest.java (hunk 11 above)
Comment (AndreasMager) [reply osgi_test/src/test/java/io/moquette/test/osgi/MoquetteOsgiTest.java]: I will debug it further.

**Comment by andsel** [comment]
@AndreasMager could you fix it and keep separated the gradle part from the fixes pushed into the PR?

**Comment by AndreasMager** [comment]
@andsel  Yes #349 & #350 contains all former code changes. I will rebase this branch on top both and see if the ogsi stuff works well.

File: broker/build.gradle (hunk 2 above)
Comment (AndreasMager) [reply broker/build.gradle]: Good hint.

File: build.gradle (hunk 1 above)
Comment (AndreasMager) [reply build.gradle]: removed

File: build.gradle (hunk 6 above)
Comment (AndreasMager) [reply build.gradle]: A flavor

Comment (AndreasMager) [reply build.gradle]: These plugins will add inline 100 loc.

**Comment by AndreasMager** [comment]
@andsel the branch is squashed and all your wishes/markers are solved.
The osgi test setup code has changed a little. The Maven build system adds information which can be read by ops4j.pax.exam version fetch script. I hope they will add a gradle version in the future.

**Comment by andsel** [comment]
Thanks a lot @AndreasMager PR integrated
